src/main/core/work/event.rs

```diff
@@ -61,105 +66,121 @@ impl PartialEq for Event {
+    src_host_event_id: u64,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct LocalEventData {
+    task: TaskRef,
+    event_id: u64,
+}
+
+impl PacketEventData {
+    /// Get the packet contained within this event.
+    pub fn packet(self) -> Packet {
+        self.packet
+    }
+}
+
+impl PartialOrd for PacketEventData {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        // sort by src host ID, then by event ID
         let cmp = self
-            .time
-            .cmp(&other.time)
-            .then_with(|| self.dst_host_id.cmp(&other.dst_host_id))
-            .then_with(|| self.src_host_id.cmp(&other.src_host_id))
+            .src_host_id
+            .cmp(&other.src_host_id)
             .then_with(|| self.src_host_event_id.cmp(&other.src_host_event_id));
 
         // if the above fields were all equal (this should ideally not occur in practice since it
         // leads to non-determinism, but we handle it anyways)
         if cmp == std::cmp::Ordering::Equal {
-            if self.task != other.task {
-                // tasks are not equal, so the events must not be equal
+            if self.packet != other.packet {
+                // packets are not equal, so the events must not be equal
                 assert_ne!(self, other);
                 // we have nothing left to order them by
                 return None;
             }
 
-            // tasks are equal, so the events must be equal
+            // packets are equal, so the events must be equal
             assert_eq!(self, other);
         }
 
         Some(cmp)
     }
 }
 
-mod export {
-    use super::*;
-
-    use shadow_shim_helper_rs::simulation_time::{CSimulationTime, SimulationTime};
-
-    #[no_mangle]
-    pub unsafe extern "C" fn event_new(
-        task_ref: *mut TaskRef,
-        time: CSimulationTime,
-        src_host: *const Host,
-        dst_host_id: HostId,
-    ) -> *mut Event {
-        let task_ref = unsafe { task_ref.as_mut() }.unwrap();
-        let src_host = unsafe { src_host.as_ref().unwrap() };
-        let time = EmulatedTime::from_abs_simtime(SimulationTime::from_c_simtime(time).unwrap());
-
-        Box::into_raw(Box::new(Event::new(
-            task_ref.clone(),
-            time,
-            src_host,
-            dst_host_id,
-        )))
-    }
-
-    #[no_mangle]
-    pub unsafe extern "C" fn event_free(event: *mut Event) {
-        assert!(!event.is_null());
-        unsafe { Box::from_raw(event) };
-    }
-
-    /// Execute the event. **This frees the event.**
-    #[no_mangle]
-    pub unsafe extern "C" fn event_executeAndFree(event: *mut Event, host: *const Host) {
-        assert!(!event.is_null());
-        let event = unsafe { Box::from_raw(event) };
-        let host = unsafe { host.as_ref().unwrap() };
-
-        event.execute(host);
+impl LocalEventData {
+    /// Get the task contained within this event.
+    pub fn task(self) -> TaskRef {
+        self.task
     }
+}
```

Comment:
Implement the `From` trait here instead?

Reply:
I've made it `impl From<LocalEventData> for TaskRef` instead.